dht_bootstrap.go

```diff
@@ -163,22 +109,68 @@ func (dht *IpfsDHT) runBootstrap(ctx context.Context, cfg BootstrapConfig) error
 		return err
 	}
 
-	// Do all but one of the bootstrap queries as random walks.
-	for i := 0; i < cfg.Queries; i++ {
-		err := doQuery(i, "random ID", dht.randomWalk)
-		if err != nil {
-			return err
+	buckets := dht.routingTable.GetAllBuckets()
+	var wg sync.WaitGroup
+	errChan := make(chan error)
+
+	for bucketID, bucket := range buckets {
+		if time.Since(bucket.RefreshedAt()) > dht.bootstrapCfg.BucketPeriod {
+			wg.Add(1)
+			go func(bucketID int, errChan chan<- error) {
+				defer wg.Done()
+				// gen rand peer in the bucket
+				randPeerInBucket := dht.routingTable.GenRandPeerID(bucketID)
+
+				// walk to the generated peer
+				walkFnc := func(c context.Context) error {
+					_, err := dht.FindPeer(ctx, randPeerInBucket)
+					if err == routing.ErrNotFound {
+						return nil
+					}
+					return err
+				}
+
+				if err := doQuery(bucketID, randPeerInBucket.String(), walkFnc); err != nil {
+					errChan <- errors.Wrapf(err, "failed to do a random walk on bucket %d", bucketID)
+				}
+			}(bucketID, errChan)
 		}
 	}
 
-	// Find self to distribute peer info to our neighbors.
-	return doQuery(cfg.Queries, fmt.Sprintf("self: %s", dht.self), dht.selfWalk)
+	// wait for all walks to finish & close the error channel
+	go func() {
+		wg.Wait()
+		close(errChan)
+	}()
+
+	// accumulate errors from all go-routines. ensures wait group is completed by reading errChan until closure.
+	var errStrings []string
+	for err := range errChan {
+		errStrings = append(errStrings, err.Error())
+	}
+	if len(errStrings) == 0 {
+		return nil
+	} else {
+		return fmt.Errorf("errors encountered while running bootstrap on RT:\n%s", strings.Join(errStrings, "\n"))
+	}
 }
 
-func (dht *IpfsDHT) BootstrapRandom(ctx context.Context) error {
-	return dht.randomWalk(ctx)
+// Traverse the DHT toward the self ID
+func (dht *IpfsDHT) selfWalk(ctx context.Context) error {
+	queryCtx, cancel := context.WithTimeout(ctx, dht.bootstrapCfg.Timeout)
+	defer cancel()
+	_, err := dht.FindPeer(queryCtx, dht.self)
+	if err == routing.ErrNotFound {
+		return nil
+	}
+	return err
 }
 
-func (dht *IpfsDHT) BootstrapSelf(ctx context.Context) error {
```

Comment:
Are we OK simply dropping methods from the public interface, @Stebalien? We usually frown upon this. Couldn't these methods be turned into shims, and marked Deprecated?

Reply:
We try to avoid silently breaking APIs or breaking APIs on a whim (e.g., preference, style, etc.). In thi scase:

1. This function was recently added.
2. We completely revamped the bootstrap system anyways and had to make breaking changes to the bootstrap config (https://github.com/libp2p/go-libp2p-kad-dht/pull/384#discussion_r315959397).

On the other hand, you're right. This change wasn't completely necessary and probably should have gone through a deprecation period.